File: typescript/src/message.ts
/**
 * The undecoded payload a channel broadcast or transmitted. Its shape is
 * entirely up to the channel, so `json()` it into the expected type.
 *
 * Go keeps the server's bytes verbatim in a `json.RawMessage`. Here `raw` is
 * the payload re-encoded from the frame the protocol decoded, so the text is
 * canonical JSON rather than byte-for-byte what arrived — the value is the
 * same, the whitespace may not be.
 */
export class Message {
  constructor(readonly raw: string) {}

  /** Decodes the payload. */
  json<T = unknown>(): T {
    return JSON.parse(this.raw) as T;
  }

  toString(): string {
    return this.raw;
  }

  toJSON(): unknown {
    return this.json();
  }
}

File: swift/Sources/ActionCable/Message.swift
import Foundation

/// The undecoded payload a channel broadcast or transmitted. Its shape is
/// entirely up to the channel, so `decode` it into the expected type.
public struct Message: Sendable, Hashable {
    public let data: Data

    public init(_ data: Data) {
        self.data = data
    }

    public func decode<T: Decodable>(_ type: T.Type = T.self, using decoder: JSONDecoder = JSONDecoder()) throws -> T {
        try decoder.decode(type, from: data)
    }

    public var text: String {
        String(decoding: data, as: UTF8.self)
    }
}

extension Message: CustomStringConvertible {
    public var description: String {
        text
    }
}
